### PR TITLE
Advance Payment Patch

### PR DESCRIPTION
**Pull request by Olawale1**
Hi

Trust you're doing great. I've added all the basic components for the Advance Payment functionality

Changes to PY file:

1. Added whitelist method for creating Advance Payment entry 
2. Added whitelist method for calculating Total Advance Paid 

Changes to JS file:

1. Added script and made relevant changes for Advance Payment button
2. Added script for calling Total Advance Paid method
3. Added script to auto-fetch Default Advance Account from Company master

Changes to JSON file:

1. Added field for Default Advance Account in Company master
2. Added field for Advance Payment Required in Expense Claim form
3. Added field for Total Advance Paid in Expense Claim form

Changes to Documentation

1. Added 'Managing Advance Payments' section to the Expense Claim documentation

Screenshots of End Results:

company_advance_account

advance_payment_required

make_advance_payment

advance payment patch2

advance payment patch3

Thanks

Kind regards,

### CODE REVIEW

**Comment by rmehta** [comment]
@Olawale1 can you check the codacy issue?

**Comment by Olawale1** [comment]
There's only 1 Codacy error left (no unused vars) but the script needs that line in the code

Thanks

**Comment by rmehta** [comment]
Just remove the `var doc =` part and let the sync remain

**Comment by Olawale1** [comment]
Thanks a lot @rmehta 

I've fixed it :)

File: erpnext/hr/doctype/expense_claim/expense_claim.js [hunk 1]
```diff
@@ -152,10 +172,30 @@ erpnext.expense_claim = {
 };
 
 frappe.ui.form.on("Expense Claim", {
+	refresh: function(frm) {
+		var doc = frm.doc;
+		if(doc.advance_account)
+			frappe.call({
+				method: "erpnext.hr.doctype.expense_claim.expense_claim.update_advance_paid",
+				args: {
+					"docname": cur_frm.doc.name,
```
Comment (mbauskar) [inline erpnext/hr/doctype/expense_claim/expense_claim.js]: no need to use the `cur_frm.doc` just use the `doc.name`

File: erpnext/hr/doctype/expense_claim/expense_claim.js [hunk 2]
```diff
@@ -152,10 +172,30 @@ erpnext.expense_claim = {
...
+					"employee": doc.employee,
+					"advance_account": doc.advance_account
+				},
+				callback: function(r) {
+					cur_frm.set_value("total_advance_paid", r.message.amt);
```
Comment (mbauskar) [inline erpnext/hr/doctype/expense_claim/expense_claim.js]: `cur_frm` will deprecated, please use the `frm` instead

**Review by mbauskar — CHANGES_REQUESTED** [review]
@Olawale1,

Thanks for the PR,

Please see my comments, Also can you please add simple tests cases ?

File: erpnext/hr/doctype/expense_claim/expense_claim.js (hunk 2 above)
Comment (Olawale1) [inline erpnext/hr/doctype/expense_claim/expense_claim.js]: Hi @mbauskar 

Thanks for the helpful review. Not sure how to add Test Cases. Could you give some guidance on this?

Kind regards,

**Comment by Olawale1** [comment]
Hi

I've updated the JS code by removing `cur_frm` which is soon to be deprecated. Would appreciate assistance with Test Cases

Thanks

**Comment by mhbu50** [comment]
Good Job @Olawale1  
you can do same as this : 
https://github.com/frappe/erpnext/blob/c54510660e8c7be232862f8175334c40de2e65df/erpnext/schools/doctype/instructor/test_instructor.js

**Comment by Olawale1** [comment]
Hi @mhbu50 

Sorry I missed replying this. Thanks for your encouragement and assistance :)

I've checked the reference but don't see how to adapt that for this PR. I'll require some more detailed guidance for this

BTW if the purpose is to confirm the feature works, won't it suffice to use screenshots or recording?

Thanks

**Comment by tundebabzy** [comment]
No. The videos might prove that your code works right but it won't tell us when code in another part of the software breaks your code later.

**Comment by Olawale1** [comment]
Oh, I see... makes sense. In that case, this should actually be a kind of automated process

Still require some assistance here

Thanks

**Comment by rmehta** [comment]
@nabinhait please clean and merge this...

**Comment by Olawale1** [comment]
Thanks plenty

**Comment by Olawale1** [comment]
Thanks a lot @nabinhait

**Comment by tundebabzy** [comment]
#10632 reverted

**Pull request closed without merging**